### PR TITLE
convert util2.c to D

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: src/dmd/backend/util2.d [hunk 1]
```diff
@@ -51,15 +44,15 @@ void file_progress()
...
 {
     fflush(stdout);
     printInternalFailure(stdout);
     printf("Internal error: %s %d\n",file,line);
     err_exit();
-#if __clang__
-    __builtin_unreachable();
-#endif
+//#if __clang__
+//    __builtin_unreachable();
```
Comment (ibuclaw) [inline src/dmd/backend/util2.d]: `assert(0)` would be the equivalent here.

**Review by ibuclaw — APPROVED** [review]
There's a lack of indentation in versioned code, but nothing that really sticks out as being wrong.

**Comment by WalterBright** [comment]
I try to fix indenting in a separate PR all its own, in order to minimize diffs in the semantic PRs.

**Comment by ibuclaw** [comment]
Why? It's possible to review PRs with whitespace changes hidden to see instead what content changed.

**Comment by thewilsonator** [comment]
Appveyor:
```
LINK : warning LNK4044: unrecognized option '/delexe/la'; ignored
backend.lib(out.obj) : error LNK2019: unresolved external symbol "void __cdecl util_free(void *)" (?util_free@@YAXPEAX@Z) referenced in function "void __cdecl writefunc2(struct Symbol *)" (?writefunc2@@YAXPEAUSymbol@@@Z)
backend.lib(cgcod.obj) : error LNK2001: unresolved external symbol "void __cdecl util_free(void *)" (?util_free@@YAXPEAX@Z)
backend.lib(cgcv.obj) : error LNK2001: unresolved external symbol "void __cdecl util_free(void *)" (?util_free@@YAXPEAX@Z)
backend.lib(go.obj) : error LNK2001: unresolved external symbol "void __cdecl util_free(void *)" (?util_free@@YAXPEAX@Z)
backend.lib(util2.obj) : error LNK2019: unresolved external symbol controlc_saw referenced in function controlc_handler
backend.lib(cgcod.obj) : error LNK2019: unresolved external symbol "void * __cdecl util_calloc(unsigned int,unsigned int)" (?util_calloc@@YAPEAXII@Z) referenced in function "void __cdecl codgen(struct Symbol *)" (?codgen@@YAXPEAUSymbol@@@Z)
backend.lib(cgcod.obj) : error LNK2019: unresolved external symbol "void * __cdecl util_realloc(void *,unsigned int,unsigned int)" (?util_realloc@@YAPEAXPEAXII@Z) referenced in function "void __cdecl cse_save(struct CodeBuilder &,unsigned int)" (?cse_save@@YAXAEAUCodeBuilder@@I@Z)
backend.lib(cgcv.obj) : error LNK2001: unresolved external symbol "void * __cdecl util_realloc(void *,unsigned int,unsigned int)" (?util_realloc@@YAPEAXPEAXII@Z)
backend.lib(varstats.obj) : error LNK2001: unresolved external symbol "void * __cdecl util_realloc(void *,unsigned int,unsigned int)" (?util_realloc@@YAPEAXPEAXII@Z)
..\generated\windows\release\64\dmd.exe : fatal error LNK1120: 4 unresolved externals
Error: linker exited with status 1120
```

File: src/dmd/backend/util2.d [hunk 2]
```diff
@@ -90,16 +81,16 @@ void util_exit(int exitcode)
     exit(exitcode);                     /* terminate abnormally         */
 }
 
+version (Windows)
```
Comment (rainers) [inline src/dmd/backend/util2.d]: use CRuntime_DigitalMars instead.

File: src/dmd/backend/util2.d [hunk 3]
```diff
@@ -282,108 +282,137 @@ int binary(const char *p, size_t len, const char ** table, int high)
...
+enum UTIL_PH = true;
 
-#if _WIN32
-void *util_malloc(unsigned n,unsigned size)
+version (MEM_DEBUG)
+    enum MEM_DEBUG = false; //true;
+else
+    enum MEM_DEBUG = false;
+
+version (Win32)
```
Comment (rainers) [inline src/dmd/backend/util2.d]: Should be `Windows`  as `_WIN32` is defined for x64 aswell.

**Comment by WalterBright** [comment]
> It's possible to review PRs with whitespace changes hidden to see instead what content changed.

1. I use other diff programs as well to compare versions, and they don't.
2. The github one only works with leading whitespace. Doing a proper reformat requires more, like lining up the comments again.
3. Again, and I want to beat this dead horse that keeps rising again, translating from one language to another is best done with MINIMAL DIFFS. I really do mean that. MINIMAL DIFFS. Can I say it again? MINIMAL DIFFS. That means no refactoring, reformatting, or fixing. I don't care how tempting it is to do that. I have a lot of experience translating code. This methodology results in the LEAST amount of work in the end. And if a bug is introduced in the translation, and this is inevitable, it is MUCH EASIER to find if the diffs are minimal.

For example, most of the line diffs are are -> being replaced with . So when I'm looking for the bug I introduced in the translation, I copy the .c file to tmp, and in tmp, I replace all the -> with dots. Voila, then the real differences stand out. This is what I mean when I use other diff programs.

File: src/dmd/backend/util2.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/backend/util2.d]: Thanks, I had forgotten about that.

**Comment by WalterBright** [comment]
BTW, the right way to fix formatting is with a PR that does nothing else but fix formatting. That makes it easy for the reviewer. Having something else altered in the middle of a large formatting PR is a recipe for a lot of wasted time.

**Comment by thewilsonator** [comment]
There's still something funky going on. Looks like redefinition problems?

**Comment by WalterBright** [comment]
Who knows. Not much of a clue from the log.

**Comment by thewilsonator** [comment]
Can I suggest taking it off the auto merge queue for the time being?

**Comment by WalterBright** [comment]
This PR is a marvelous example of why minimizing diffs is helpful :-(

**Comment by WalterBright** [comment]
@rainers apparently the failure is due to the inline assembler version of `binary`. How do I use versions to distinguish that LDC built dmd?

File: src/dmd/backend/util2.d [hunk 4]
```diff
@@ -123,105 +127,107 @@ void _STD_controlc()
...
  *      tab     array of pointers to strings
  *      n =     number of pointers in the array
  * Returns:
  *      index (0..n-1) into tab[] if we found a string match
  *      else -1
  */
 
-#if TX86 && __DMC__ && !_DEBUG_TRACE
+version (X86) version (CRuntime_DigitalMars)
+    version = X86asm;
```
Comment (thewilsonator) [inline src/dmd/backend/util2.d]: @WalterBright 
```d
version(LDC) { }
else
version (X86) version (CRuntime_DigitalMars
    version = X86asm;
```

should do the trick.

File: src/dmd/backend/util2.d [hunk 5]
```diff
@@ -51,15 +44,15 @@ void file_progress()
  * Alternative assert failure.
  */
 
-void util_assert(const char *file, int line)
+void util_assert(const(char)* file, int line)
 {
     fflush(stdout);
-    printInternalFailure(stdout);
+//    printInternalFailure(stdout);
```
Comment (rainers) [inline src/dmd/backend/util2.d]: Why is `printInternalFailure` removed?

File: src/dmd/backend/util2.d [hunk 6]
```diff
@@ -109,12 +102,23 @@ static void __cdecl controlc_handler(void)
  * Trap control C interrupts.
  */
 
-#if !MARS
+version (MARS) { } else
+{
+
+version (CRuntime_DigitalMars)
```
Comment (rainers) [inline src/dmd/backend/util2.d]: This duplicates version CRuntime_DigitalMars above.

File: src/dmd/backend/util2.d (hunk 5 above)
Comment (WalterBright) [reply src/dmd/backend/util2.d]: oops

File: src/dmd/backend/util2.d (hunk 6 above)
Comment (WalterBright) [reply src/dmd/backend/util2.d]: ok

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**